drift-transport-netty/src/main/java/com/facebook/drift/transport/netty/codec/HeaderTransport.java

```diff
@@ -262,10 +338,21 @@ private static int readVariableLengthInt(ByteBuf messageHeader)
                 return Optional.empty();
             }
 
-            byte protocolId = buffer.getByte(buffer.readerIndex());
+            messageHeader = buffer.readBytes(headerSize);
+            byte protocolId = messageHeader.readByte();
             Protocol protocol = Protocol.getProtocolByHeaderTransportId(protocolId);
 
-            buffer.skipBytes(headerSize);
+            int numberOfTransforms = readVariableLengthInt(messageHeader);
+            checkArgument(numberOfTransforms >= 0, "invalid number of transforms");
+            ImmutableList.Builder<Transform> readTransformBuilder = builder();
+
+            for (int i = 0; i < numberOfTransforms; i++) {
+                int transId = readVariableLengthInt(messageHeader);
+                readTransformBuilder.add(Transform.fromId(transId));
+            }
+            if (numberOfTransforms > 0) {
+                buffer = untransform(buffer, readTransformBuilder.build());
```

Comment:
This buffer is not guaranteed to contain the whole message buffered. It may only contain a few bytes. In this case the `ZlibTransformer#untransform` will fail with `Unexpected end of ZLIB input stream`.

For example the `TestHeaderTransport` fails when the message becomes larger: https://gist.github.com/arhimondr/73e485aedb66b5fcf53a86335b4278da

Most likely you would need to introduce a `untransformPartial` method, that can do the best on decompressing a partial message. 

To implement one, most likely you would need to use the `Inflater` interface (see `Inflater#setInput`, `Inflater#inflate`). This code will be quite sketchy though. Thus make sure it is well tested.

Reply:
`tryDecodeFrameInfo` will be called if the frame is too long. But reading the code here:
```
            pipeline.addLast("frameEncoder", new LengthFieldPrepender(Integer.BYTES));
            pipeline.addLast("thriftFramedDecoder", new ThriftFramedDecoder((bufAllocator, input) -> HeaderTransport.tryDecodeFrameInfo(bufAllocator, input), toIntExact(maxFrameSize.toBytes())));
            pipeline.addLast(new HeaderCodec());
``` 
It seems like we should always decode a bytebuf to contain a full frame? Otherwise, `HeaderCodec` will fail to decode big messages as well. 

However, I added support to try decode. I can expand on it. But I am a little worried for cases like `Encryption` we won't be able to decode partially.

Reply:
> It seems like we should always decode a bytebuf to contain a full frame?

From the message only the method name and the sequence id is needed. Those are the first thrift fields, and can be decoded without buffering of the entire message

> But I am a little worried for cases like Encryption we won't be able to decode partially

Usually encryption also happens in blocks (similar to compression), so we should be able to decode it partially as well.